cloudevents/working-drafts/xml-format.md

```diff
@@ -0,0 +1,367 @@
+structure and metadata description of events. This specification defines how
+elements defined in the CloudEvents specification are to be represented using
+[Extensible Markup Language (XML)](xml-spec) documents.
+
+* The [Attributes](#2-attributes) section describes the representation and
+data type mappings for CloudEvents context attributes.
+
+* The [Data](#3-data) section defines the container for the data portion of a
+CloudEvent.
+
+* The [Envelope](#4-envelope) section defines an XML element for CloudEvents
+attributes and an associated media type.
+
+* The [Batch](#5-xml-batch-format) section describes how multiple CloudEvents
+can be packaged into a single XML document.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Approach
+
+The XML representation used is deliberately verbose in favor of readability
+and deterministic processing.
+
+Preservation of type information for CloudEvent attributes is supported allowing
+custom extensions to be communicated without type loss.
+
+A schema-less approach has been taken favoring convention over rigid document
+structure.
+
+The namespace `http://cloudevents.io/xmlformat/V1` MUST be used.
+
+When namespace prefixes are used a prefix of `ce` is preferred but MUST NOT
+be expected from an XML document processing perspective.
+
+An XML document preamble SHOULD be included to ensure deterministic processing.
+
+## 2. Attributes
+
+The CloudEvents type system is mapped to a set of CloudEvent specific
+type designators as follows :
+
+| CloudEvents Type  |  XML Format Type | XML Schema Type | Notes |
+| :-----------------| :----------------| :--------------- | :---- |
+| Boolean       | ce:boolean | [xs:boolean][xml-primitives] | `true` or `false` |
+| Integer       | ce:int | [xs:int][xml-primitives] | |
+| String        | ce:string | [xs:string][xml-primitives] | |
+| Binary        | ce:binary | [xs:base64Binary][xml-primitives] |  |
+| URI           | ce:uri | [xs:anyURI][xml-primitives] following [RFC 3986][rfc3986]| |
+| URI-reference | ce:uriRef | [xs:anyURI][xml-primitives] following [RFC 3986][rfc3986] | |
+| Timestamp     | ce:timestamp | [xs:dateTime][xml-primitives] following [RFC 3339][rfc3339] | |
+
+Each CloudEvent context attribute is represented as an XML element whose local
+name exactly matches that of the attribute.
+
+See the [envelope](#4-envelope) for special handing of the `specversion` context attribute.
+
+Extension attributes MUST be decorated with the appropriate CloudEvent format
+designators using using the `xsi:type` XML attribute, this allows them to be exchanged
+without loss of type information.
+
+REQUIRED and OPTIONAL context attributes SHOULD NOT be decorated with an `xsi:type`.
+
+No other XML element attributes are expected, if present they MUST be ignored during
```

Comment:
I'm confused by the phrase "no other". We have spec-defined and we have "extensions", what else is there? Anything not defined by the spec must be an extension and ignoring it doesn't seem right.

Reply:
Spec-defined and extensions are *CloudEvent* attributes. These are *XML* attributes.

So an attribute like this:

```xml
<ce:event>
  <ce:id inquisition="Spanish" />
</ce:event>
```

No-one expects that. Personally I think they should cause a parsing error though.

Reply:
Ah, i totally misread that as just "XML elements" - thanks

Reply:
@jskeet - I got the Monty Python reference there .. well played sir.

Reply:
Is there a reason we don't want to allow for extensions (e.g. other attributes on the xml elements)? There's no guarantee that the receiver will understand them, but extensions are really useful in general

Reply:
I think it depends on what the expectations are. Personally I *tend* to favour really strict specs, resulting in a relatively low but very-common denominator, rather than a more open-ended spec which leaves a lot of room for some implementations to do one thing and others to do another, resulting in a feature which no-one can use confidently.

Reply:
Just seems like a shame for someone to not be allowed to use a feature of the format just because they're sending a CloudEvent. I view this along the lines of a Kube based project saying "we're not going to allow people to add new annotations and labels to the CR for the CRD that we define". It makes it harder for third parties to build on top of what's there w/o a lot of extra work and sync'ing. 

All I'm looking for is to replace the "MUST" here with something like:
```
This specification allows for additional non-CloudEvents namespaced XML elements and XML element
attributes to appear. However, this specification makes no requirement on receivers to understand
those extensions.
```
Note: I have the same feelings about extension elements. This allows people to build on top of our stuff w/o a lot of pain - and, what's interesting to me, is that this actually gives people an incentive to use XML since I'm not sure what other formats give them this level of flexibility.

First we resurrect XML, then we bring back SOAP! :-)

Reply:
Without extension attributes existing, I'd agree. But they *do* exist for this purpose - they're the way we have extensibility within some guardrails. I'd suggest the incentive to use XML should be more about fitting in with other bits of infrastructure (e.g. including CloudEvents within a document that might contain other information) rather than relying on additional capabilities.

We could make this strictness opt-out, in terms of "processors MUST fail by default if unexpected attributes, elements or text nodes are present, but MAY provide mechanisms of parsing less strictly".

Reply:
I probably wasn't clear. I meant there's no easy way to add extra metadata to each CE extension property in formats like structured JSON w/o doing something funky like add another CE extension. With XML you can add that metadata directly on the new CE extension xml element as an xml attributes. And that's pretty cool/handy.

Likewise, structured JSON doesn't allow you to add extra properties w/o them being interpreted as CE extension properties. But you can with XML via namespaces. Again, cool. Worth pointing out that HTTP CE binary mode does let you do this via non CE- prefixed HTTP headers.

w.r.t. text, what about:
```
Receivers MAY reject messages with extra XML element attributes or XML elements
if they choose to do so.
```
I wouldn't want to default to failing since that's a high bar to force everything look for extra things rather than blindly ignore them.

Reply:
> I wouldn't want to default to failing since that's a high bar to force everything look for extra things rather than blindly ignore them.

I'd argue it gives the best developer experience though - if your CloudEvent is parseable with one SDK, it should be parseable by all, on the default settings.

I've never been comfortable with Postel's Law - my experience is that that's the route to everyone having to work around laxness in other systems, because forcing people to actually do the right thing is too hard. ("But it works on my machine.")

Something to discuss in the next WG meeting?

Reply:
This sounds like a discussion item.